### PR TITLE
Monolithic Update to DevOps

### LINKED ISSUE #111: Blacklisted cmake 3.18.2 because macos build failed
See [commit abff69f](https://github.com/WildbookOrg/wildbook-ia/commit/abff69f619215947eccb7972326e599e7e4eec69).

 Don't use cmake 3.18.2 when building wheel

Our macos build has been failing since 2020-08-21:

```
+ pip wheel . -w /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheelvcmdx5mi/built_wheel --no-deps
Processing /Users/runner/work/wildbook-ia/wildbook-ia
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /private/tmp/cibw_bin/python /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-srqflxar/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=42' wheel 'setuptools_scm[toml]>=3.4' scikit-build cmake ninja
       cwd: None
  Complete output (23 lines):
  Collecting setuptools>=42
    Downloading setuptools-49.6.0-py3-none-any.whl (803 kB)
  Collecting wheel
    Downloading wheel-0.35.1-py2.py3-none-any.whl (33 kB)
  Collecting setuptools_scm[toml]>=3.4
    Downloading setuptools_scm-4.1.2-py2.py3-none-any.whl (27 kB)
  Collecting scikit-build
    Downloading scikit_build-0.11.1-py2.py3-none-any.whl (72 kB)
  Collecting cmake
    Downloading cmake-3.18.2.tar.gz (28 kB)
      ERROR: Command errored out with exit status 1:
       command: /private/tmp/cibw_bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-iy6chcw0/cmake/setup.py'"'"'; __file__='"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-iy6chcw0/cmake/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-pip-egg-info-4z_omqj3
           cwd: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-iy6chcw0/cmake/
      Complete output (5 lines):
      Traceback (most recent call last):
        File "<string>", line 1, in <module>
        File "/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-iy6chcw0/cmake/setup.py", line 7, in <module>
          from skbuild import setup
      ModuleNotFoundError: No module named 'skbuild'
      ----------------------------------------
```

cmake 3.18.2 was released on 2020-08-21 and that appears to break the macos
build.

We might need to look into this so raising an issue.

### PR DESCRIPTION
**Pull request by bluemellophone**
This PR does several updates to how our development Docker images are built using the path `wbia-base` to `wbia-dependencies` to `wbia-provision` to `wbia`.  Here are the following major updates:
- Refactored `devops/[base|dependency|provision]/Dockerfile` to bring them in line with production build versions on private Azure DevOps fork and its `next` branch.
- Refactored `devops/Dockerfile` to bring optimizations from runtime version in `Dockerfile` by adding `gosu` and a new `devops/_config/entrypoint.sh`
- Added a tagged Docker image build and publish GitHub Action
- Added smoke tests for `wbia`, `wbia-utool`, `wbia-vtool`, all of the `wbia-tpl-*` packages, `wbia-cnn`, and all production-ready `wbia-plugin-*` packages
- Re-added a `git pull` to the `wbia` image build to refresh all code versions cloned from `wbia-provision` before deleting the `.git` folders.
- Bumps Docker image build Python version from 3.6 to 3.7
- Bumps Docker image built OpenCV version from 3.4.10 to 3.4.11
- Downgrade and pegged to TensorFlow 1.15.3 for `wbia-plugin-pie`

For known issues, this PR is building on Ubuntu and macOS and all tests are passing.  It also:
- Resolves issue #118 since it is loading the dynamic library during the smoke tests, ref: https://github.com/WildbookOrg/wildbook-ia/runs/1132428363?check_suite_focus=true#step:5:312
- Resolves the final two items in #107 by pulling over the required `apt` dependencies and `cmake` configuration from `wbia-pypkg-build` for OpenCV.  The `apt` dependencies to build OpenCV are now simply `libgflags-dev`, `libeigen3-dev`, `libgtk2.0-dev`, and `numpy`, ref: https://github.com/WildbookOrg/wildbook-ia/blob/build-process-merge/devops/dependencies/Dockerfile#L20
- Resolves #111 by adding an explicit `cmake!=3.18.2` in the pip install within `devops/base/Dockerfile`, ref: https://github.com/WildbookOrg/wildbook-ia/blob/build-process-merge/devops/base/Dockerfile#L87

For hot-fixes in production, this PR:
- Adds background job retries on failure (3 retries with 60 second delay)
- Adds DependencyCache `sqlite3.IntegrityError` fix for parallel jobs writing to the same rows (3 retries with 5 second delay)
- Add import errors from updated dependencies (exposed by documentation build)
- Added `nvtop` inside the container for GPU debug to mirror the `htop` already installed from `apt` for CPU debug

For cleanliness, this PR also:
- Linted and fixed PEP8 errors for the entire codebase
- Bumped software version to 3.3.0 for documentation purposes
- Fixed references to deleted modules and re-built documentation
- Remove remaining \_\_future\_\_ imports
- Add file handling closing where exposed by testing
- Added Docker image labels via http://label-schema.org/rc1/
- Added "root by default" when no data volumes are mounted, e.g. running simply `docker run wildme/wbia:develop`

To put a resolving achievement on this PR, we significantly reduced the size of the `wildme/wbia` image by drastically reducing the `apt` packages that are installed and being more diligent about constructing our final layers out of the multi-stage Docker build process.  In summary, our `apt` packages are now simply:

```
# Install build requirements
ca-certificates
build-essential
pkg-config

# Install Python 3.7 - Pegged to Version 3.7 to support TensorFlow 1.15.3
python3.7
python3.7-dev
python3.7-gdbm
python3-pip

# Install dependencies for NVTOP
libncurses5-dev
libncursesw5-dev

# Install dependencies for OpenCV
libgflags-dev
libeigen3-dev
libgtk2.0-dev

# Install dependencies for PyHesaff
libhdf5-serial-dev

# Install dependencies for PyFLANN
liblz4-dev

# Install dependencies for pygraphviz
graphviz
libgraphviz-dev

# Install run-time tools
gosu
xvfb

# Install handy developer tools
curl
git
htop
locate
tmux
unzip
vim
wget
```

### CODE REVIEW

**Comment by karenc** [comment]
I'm reviewing changes and reformatting the commits at the same time.

**Comment by bluemellophone** [comment]
The updates today largely fix the problems of storage folder permissions and that the container could not be launched by simply executing 

```docker run wildme/wbia:develop```

Both should be fixed now.  Using a non-root user that is mapped using `HOST_USER` and `HOST_UID` allows the data directory to have the proper permissions.

**Comment by karenc** [comment]
@bluemellophone I was reformatting the commits without `pre-commit` turned on.  Then I tried doing `pre-commit run -a`.  I was expecting brunette to not change any of the files but it modified more than 100 files:

```diff
diff --git a/_dev/super_setup_old.py b/_dev/super_setup_old.py
index 7c18e6cb4..2cbf091cf 100755
--- a/_dev/super_setup_old.py
+++ b/_dev/super_setup_old.py
@@ -1036,13 +1036,7 @@ def move_wildme(wbia_rman, fmt):
                     if origin_user not in remotes:
                         # first add a remote that is the original origin
                         origin_url = origin['url']
-                        print(
-                            '  * Create remote %r: %r'
-                            % (
-                                origin_user,
-                                origin_url,
-                            )
-                        )
+                        print('  * Create remote %r: %r' % (origin_user, origin_url,))
                         gitrepo.create_remote(origin_user, origin_url)
                     # change origin to use wildme url
                     gitorigin = gitrepo.remote('origin')
diff --git a/super_setup.py b/super_setup.py
index 3cb6e3384..1965f01fd 100755
--- a/super_setup.py
+++ b/super_setup.py
@@ -556,8 +556,7 @@ class Repo(ub.NiceRepr):
         # Print some status
         repo.debug(
             ' * branch = {} -> {}'.format(
-                repo.pygit.active_branch.name,
-                repo.pygit.active_branch.tracking_branch(),
+                repo.pygit.active_branch.name, repo.pygit.active_branch.tracking_branch(),
             )
         )
 
```

etc

Do you know what's going on?

When I went through your PEP8 changes, I found some of them to be less readable after the change:

```diff
diff --git a/wbia/constants.py b/wbia/constants.py
index 563cd54b8..de3dbfcb8 100644
--- a/wbia/constants.py
+++ b/wbia/constants.py
@@ -32,76 +32,28 @@ TAU = 2.0 * PI
 
 VIEWTEXT_TO_YAW_RADIANS = OrderedDict(
     [
-        (
-            'right',
-            0.000 * TAU,
-        ),
-        (
-            'frontright',
-            0.125 * TAU,
-        ),
-        (
-            'front',
-            0.250 * TAU,
-        ),
-        (
-            'frontleft',
-            0.375 * TAU,
-        ),
-        (
-            'left',
-            0.500 * TAU,
-        ),
-        (
-            'backleft',
-            0.625 * TAU,
-        ),
-        (
-            'back',
-            0.750 * TAU,
-        ),
-        (
-            'backright',
-            0.875 * TAU,
-        ),
+        ('right', 0.000 * TAU,),
+        ('frontright', 0.125 * TAU,),
+        ('front', 0.250 * TAU,),
+        ('frontleft', 0.375 * TAU,),
+        ('left', 0.500 * TAU,),
+        ('backleft', 0.625 * TAU,),
+        ('back', 0.750 * TAU,),
+        ('backright', 0.875 * TAU,),
     ]
 )
 
```

I think it might be good to revert them anyway.

**Comment by bluemellophone** [comment]
We can employ the brunette/black macro `#fmt: on/off`.  Ref: https://github.com/psf/black/commit/0677a539370b296399854e427ce7df2955ecfe57

**Comment by karenc** [comment]
What do you mean? Is it ok for me to revert some of your PEP8 changes by running `pre-commit`? I would prefer to do that because some of the changes made the code less readable.

File: .github/workflows/docker-publish.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      # Log into container registries
+      - name: Log into Docker Hub
+        run: echo "${{ secrets.WBIA_WILDMEBOT_DOCKER_HUB_TOKEN }}" | docker login -u wildmebot --password-stdin
+
+      - name: Log into GitHub Packages
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login docker.pkg.github.com -u ${{ github.actor }} --password-stdin
+
+      # Push containers out to container registries
+      - name: Push to GitHub Packages
+        if: ${{ github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags/v') }}
```
Comment (karenc) [inline .github/workflows/docker-publish.yaml:37]: This file looks almost the same as nightly.  Why don't we just add this to nightly?

**Comment by karenc** [comment]
This PR was not ready.  I specifically marked this PR as draft because I was in the middle of reviewing it and trying to make sense of all the changes by splitting them into reasonable commits.  The nightly build was broken.  `wbia-plugin-2d-orientation` was removed from the provision dockerfile.

@bluemellophone Please stop just merging PRs.  It would have been ready tomorrow.

**Comment by bluemellophone** [comment]
T close out the discussion here, the nightly build was broken because PyDarknet had an invalid built library for CPU and a commit (https://github.com/WildbookOrg/wildbook-ia/blob/develop/devops/Dockerfile#L27) added check to ensure they loaded correctly after building.  The `wbia-plugin-2d-orientation` plugin was also intentionally removed

**Pull request merged**